blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc

```diff
@@ -0,0 +1,455 @@
+
+The big advantage is consistency and awareness of transactional boundaries,
+whereas the KStreams-based solution in its suggested form was prone to exposing intermediary aggregates.
+For instance, if you're storing a customer and three addresses, it might happen that the streaming query first creates an aggregation of the customer and the two addresses inserted first, and shortly thereafter the complete aggregate with all three addresses.
+This not the case for the approach discussed here, as you'll only ever stream complete aggregates to Kafka.
+Also this approach feels a bit more "light-weight", i.e. a simple marker annotation (together with some Jackson annotations for fine-tuning the emitted JSON structures) is enough in order to materialize aggregates from your domain model,
+whereas some more effort was needed to set up the required streams, temporary tables etc. with the KStreams solution.
+
+The downside of driving aggregations through the application layer is that it's not fully agnostic to the way you access the primary data.
+If you bypass the application, e.g. by patching data directly in the database, naturally these updates would be missed, requiring a refresh of affected aggregates.
+Although this again could be done through change data capture and Debezium:
+change events to source tables could be captured and consumed by the application itself, allowing it to re-materialize aggregates after external data changes.
+You also might argue that running JSON serializations within source transactions and storing aggregates within the source database represents some overhead.
+This often may be acceptable, though.
+
+Another question to ask is what's the advantage of using change data capture on an intermediary aggregate table over simply posting REST requests to Elasticsearch.
+The answer is the highly increased robustness and fault tolerance.
+If the Elasticsearch cluster can't be accessed for some reason, the machinery of Kafka and Kafka Connect will ensure that any change events will be propagated eventually, once the sink is up again.
+Also other consumers than Elasticsearch can subscribe to the aggregate topic, the log can be replayed from the beginning etc.
+
+Note that while we've been talking primarily about using Elasticsearch as a data sink, there are also other datastores and connectors that support complexly structured records.
+One example would be MongoDB and the https://github.com/hpgrahsl/kafka-connect-mongodb[sink connector] maintained by Hans-Peter Grahsl,
+which one could use to sink customer aggregates into MongoDB, for instance enabling efficient retrieval of a customer and all their associated data with a single primary key look-up.
+
+== Outlook
+
+The Hibernate ORM extension as well as the SMT discussed in this post can be found in our https://github.com/debezium/debezium-examples/tree/master/end-to-end-demo[examples repository].
```

Comment:
again is this the correct link? or maybe it's just not pushed yet :)

Reply:
Thx, fixed it :)